### PR TITLE
Add diagnostic for try let and try var

### PR DESCRIPTION
**Pull request by ahoppen**
Add a diagnostic if the user wrote `try let foo = bar()`. 

Also support recovery at of statements (essentially adding `allowRecovery` to `atStartOfStatement`). This required a rewrite of `atStartOfStatement`.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Tests/SwiftParserTest/translated/TryTests.swift [hunk 1]
```diff
@@ -125,71 +125,55 @@ final class TryTests: XCTestCase {
...
-      try1️⃣ 2️⃣let singleLet = foo()
+      1️⃣try let singleLet = foo()
       """,
       diagnostics: [
-        // TODO: Old parser expected error on line 1: 'try' must be placed on the initial value expression, Fix-It replacements: 1 - 5 = '', 21 - 21 = 'try '
-        DiagnosticSpec(locationMarker: "1️⃣", message: "consecutive statements on a line must be separated by ';'"),
-        DiagnosticSpec(locationMarker: "2️⃣", message: "expected expression in 'try' expression"),
-      ]
+        DiagnosticSpec(message: "'try' must be placed on the initial value expression", fixIts: ["move 'try' after '='"]),
+      ], fixedSource: "let singleLet = try foo()"
```
Comment (bnbarham) [inline Tests/SwiftParserTest/translated/TryTests.swift:132]: What happens if there's already a `try`? IIUC we'd wrap it again, ie. end up with `try try foo()`?

File: Tests/SwiftParserTest/translated/TryTests.swift [hunk 2]
```diff
@@ -204,98 +188,76 @@ final class TryTests: XCTestCase {
...
-        DiagnosticSpec(locationMarker: "1️⃣", message: "unexpected text 'try' before variable"),
-        // TODO: Old parser expected error on line 3: 'try' must be placed on the initial value expression, Fix-It replacements: 3 - 7 = '', 23 - 23 = 'try '
-        // TODO: Old parser expected error on line 3: call can throw, but errors cannot be thrown out of a property initializer
-        DiagnosticSpec(locationMarker: "2️⃣", message: "unexpected text 'try' before variable"),
-        // TODO: Old parser expected error on line 4: expected declaration
-        DiagnosticSpec(locationMarker: "3️⃣", message: "unexpected text 'try' before function"),
-      ]
+        DiagnosticSpec(locationMarker: "1️⃣", message: "'try' must be placed on the initial value expression", fixIts: ["move 'try' after '='"]),
+        DiagnosticSpec(locationMarker: "2️⃣", message: "'try' must be placed on the initial value expression", fixIts: ["move 'try' after '='"]),
+        DiagnosticSpec(locationMarker: "3️⃣", message: "'try' cannot be used with 'func'"),
```
Comment (bnbarham) [inline Tests/SwiftParserTest/translated/TryTests.swift:193]: Little bit odd, IMO a missing expression makes more sense here. If we keep this then I'd expect `try` to be unexpected and thus removed though (where as it doesn't seem to be below).

File: Tests/SwiftParserTest/translated/TryTests.swift [hunk 3]
```diff
@@ -204,98 +188,76 @@ final class TryTests: XCTestCase {
...
-      func test() throws -> Int {
-        try1️⃣ 2️⃣return
-      }
+      1️⃣try return
       """,
       diagnostics: [
-        // TODO: Old parser expected error on line 7: 'try' cannot be used with 'return'
-        DiagnosticSpec(locationMarker: "1️⃣", message: "consecutive statements on a line must be separated by ';'"),
-        DiagnosticSpec(locationMarker: "2️⃣", message: "expected expression in 'try' expression"),
+        DiagnosticSpec(message: "'try' cannot be used with 'return'"),
```
Comment (bnbarham) [inline Tests/SwiftParserTest/translated/TryTests.swift:237]: Hmmm... I'm not really sure what I'd expect here. This is what we currently do so I suppose it's fine.

This is similar to the `try while` case above. Having it expect `return try` and `while try` sort of makes more sense to me since if someone has written `try` they're probably going to call something that throws. I don't feel very strongly about this though, pretty edge case and removing also seems fine.

File: Tests/SwiftParserTest/translated/TryTests.swift (hunk 1 above)
Comment (ahoppen) [reply Tests/SwiftParserTest/translated/TryTests.swift:132]: Good catch. I changed it to not add a second `try` and added a test case for it.

File: Tests/SwiftParserTest/translated/TryTests.swift (hunk 2 above)
Comment (ahoppen) [reply Tests/SwiftParserTest/translated/TryTests.swift:193]: I think what we are trying to recover from here is that the user assumed that `try` would cover the entire variable assignment and didn't recognize that `try` can only be used on expressions. So, IMHO moving the `try` makes perfect sense to me. What would your suggestion for a diagnostic + fix-its be?

File: Tests/SwiftParserTest/translated/TryTests.swift (hunk 3 above)
Comment (ahoppen) [reply Tests/SwiftParserTest/translated/TryTests.swift:237]: Yeah, not sure if it's really worth having a diagnostic for this but since we have it now, I don't think it does any harm.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Tests/SwiftParserTest/translated/TryTests.swift (hunk 2 above)
Comment (bnbarham) [reply Tests/SwiftParserTest/translated/TryTests.swift:204]: I'm only talking about (3) in this comment. (1) and (2) seem fine to me.

For 3 I'd just expect `expected expression` after the `try`. Though if it was on the same line as the `func` I'd probably expect `unexpected try, did you mean throws?` so... 🤷

Comment (ahoppen) [reply Tests/SwiftParserTest/translated/TryTests.swift:204]: Good idea 👍
